### DISCUSSION TITLE
Using a React component using Emotion from an NPM Module breaks Gatsby SSR

### DISCUSSION THREAD

**Discussion by openscript** [emotion-js/emotion#2755] [Q&A] [answered]
Sorry for double posting, but I'm a bit desperate about this: https://github.com/gatsbyjs/gatsby/discussions/35698

As described in the post, I'm trying to use an React component inside a Gatsby project. This React component uses Emotion. Somehow the components doesn't see the Cache context and throws:

```
  1168 |   var className = "";
  1169 |   if (typeof props.className === "string") {
> 1170 |     className = getRegisteredStyles(cache.registered, registeredStyles, props.className);
       | ^
  1171 |   } else if (props.className != null) {
  1172 |     className = props.className + " ";
  1173 |   }


  WebpackError: TypeError: Cannot read properties of null (reading 'registered')
```

Any idea what I'm doing wrong? Thank you in advance!

**Comment by Andarist** [comment, marked as answer]
Most likely you have multiple copies of Emotion there

**Reply by openscript** [reply]
Thank you for your quick answer.

That is probably the case, yes. The component (https://github.com/openscript-ch/react-section-dividers) I'm using includes Emotion, as well as my Gatsby project.

So the right thing to do is externalizing Emotion (as a peer dependency) of the NPM module which provides the component?

**Reply by srmagura** [reply]
Yes, making Emotion a `peerDependency` of your library will probably fix this.

**Reply by openscript** [reply]
I've updated my library, so Emotion is a `peerDependency`, but I still run into the same error.

**Reply by openscript** [reply]
As Material UI also started using Emotion I thought I checkout the plugin which enables Material UI for Gatsby: https://github.com/hupe1980/gatsby-plugin-material-ui

Unfortunately this didn't help either. I think I'm really doing something wrong with Emotion inside my library.

**Reply by Andarist** [reply]
This definitely shouldn't specify an exact version, it should use a range (eg: `^11.9.0`)
https://github.com/openscript-ch/react-section-dividers/blob/f2e0fec045e734eeb5db25e969b1216168130ac1/package.json#L61

On top of that we can see in [the published library](https://unpkg.com/browse/react-section-dividers@0.4.5/react-section-dividers.es.js) that, despite it having Emotion as a peer dep, it **bundles** it into its own file - JavaScript runtime won't ever be able to tell that this is the same thing. It's basically making a copy of Emotion. All emotion deps should be specified as external here:
https://github.com/openscript-ch/react-section-dividers/blob/f2e0fec045e734eeb5db25e969b1216168130ac1/vite.config.ts#L15

**Reply by openscript** [reply]
Thank you for your response. I've experimented with externalizing the @emotion/react, but unfortunately this didn't help. Do I need to externalize the dependencies of `@emotion/react` as well?

**Reply by openscript** [reply]
I've externalized:
```
      external: [
        'react',
        'react-dom',
        '@emotion/cache',
        '@emotion/hash',
        '@emotion/is-prop-valid',
        '@emotion/memoize',
        '@emotion/react',
        '@emotion/serialize',
        '@emotion/styled',
        '@emotion/styled/base',
        '@emotion/unitless',
        '@emotion/utils',
      ],
```
.. but I still have the line mentioned above, which is causing the problem in my output.

**Reply by openscript** [reply]
This also looked promising https://github.com/audal/gatsby-plugin-emotion-ssg but it leads to the same error. How can I initialize this missing cache?

```
  202 |   var className = "";
  203 |   if (typeof props.className === "string") {
> 204 |     className = getRegisteredStyles(cache.registered, registeredStyles, props.className);
      |                                    ^
  205 |   } else if (props.className != null) {
  206 |     className = props.className + " ";
  207 |   }


  WebpackError: TypeError: Cannot read properties of null (reading 'registered') (from plugin: gatsby-plugin-emotion-ssg)
```

**Reply by Andarist** [reply]
I've seen such issues dozens of times - and each time it was related to running multiple versions of Emotion in an app. So that's kinda the advice I can give based on the provided information - look for Emotion copies and try to eliminate them.

I could help you if you would provide a minimal repro of the problem but without that my options are limited unless you'd like to contract me to fix this for you in your proprietary project.

**Reply by openscript** [reply]
There were multiple instances of Emotion which caused the error:
https://github.com/openscript-ch/react-section-dividers/pull/60
